### ISSUE TITLE
Candidates matching in DirectRouteControllerSelector

### ISSUE THREAD

**Issue by dawaji** [Microsoft/aspnet-api-versioning#223]
Hi,
In DirectRouteControllerSelector, line 140, when you look for matching candidates, matching is based only on declared versions on the controller :

            var matches = from candidate in directRouteCandidates
                          let controller = candidate.ActionDescriptor.ControllerDescriptor
                          where controller.GetDeclaredApiVersions().Contains( requestedVersion )
                          select controller;

In case of MapToVersion attribute it can potentially be not enough.
Let's imagine we have two different controllers, declaring the same route but mapped to different versions using the MapToVersion attribute, the DirectRouteControllerSelector would see an ambiguous route where in fact they are not.

How about matching candidates based on actions rather than controllers in the ResolveController method ?

Here is an example of what i try to achieve :

    [ApiVersion("1.0")]
    [ApiVersion("2.0")]
    [RoutePrefix("api/v{version:apiVersion}/orders")]
    public class OrdersController : ApiController
    {
        // this method is only for v1 (it is reimplemented in v2 controller)
        [MapToApiVersion("1.0")]
        [Route("test", Name = "v1", Order = 0)]
        public string Get() => "Get OrdersController v1";

        // other actions are available for both v1 and v2
    }

    [ApiVersion("2.0")]
    [RoutePrefix("api/v{version:apiVersion}/orders")]
    public class Orders2Controller : ApiController
    {
        // this method overrides the method defined in v1
        [MapToApiVersion("2.0")]
        [Route("test", Name = "v2", Order = 1)]
        public string Get() => "Get Orders2Controller";
    }

Thanks

**Comment by commonsensesoftware** [comment]
Like it or not, the default Web API routing mechanism accounts for controllers and actions. An important goal for API versioning was to not change the way routing works or have to learn anything new about routing. The implementation uses the API version as a disambiguator and nothing more. If you haven't played with ASP.NET Core yet, it's worth checking out. In _Core_, the formal concept of using a controller in routing is gone; there are only actions.

Behind the scenes, `[ApiVersion]` and `[MapToApiVersion]` have almost the same implementation. The key difference is meaning. In early, internal versions of the library they had the same name. It was ultimately changed at the action level to make it clear that actions can map, but not define API versions. It's important to understand that (in Web API) there are two resolution phase: first by controller and then by action. While this may seem limiting, it makes the rationalization of API version to controller/action significantly simpler.  The primary reason for `[MapToApiVersion]` to exist is to support version interleaving on a single type. Interleaving across types is not currently a supported scenario (and would be quite complex to implement).

It's also worth noting that trying to use inheritance this way is against the grain. Even if API versioning honored this type of interleaving, Web API does not. Although the **RouteAttribute** and **RoutePrefixAttribute** can be inherited, the [DefaultDirectRouteProvider](http://aspnetwebstack.codeplex.com/SourceControl/latest#src/System.Web.Http/Routing/DefaultDirectRouteProvider.cs) does not honor inherited attributes. Naturally, you can override the default behavior, which would likely make it work, but I don't know that it would carry over to ASP.NET Core.

I'll try not to stand too tall on the soapbox. I've now seen a few folks ask for this type of behavior. I strongly believe that inheritance is **not** the way to solve this problem; certainly not the forms that have been proposed.  Aside from being complex to support and hard to rationalize for a service team, it flat out doesn't work in all scenarios. Most of those proposing this idea are only considering an _override_ or additive scenario. APIs get sunset over time. There is no way to _disinherit_ an action.

I know there are devs out there not wanting to duplicate attribution and such, but we're at an impasse as to what is feasibly achievable. At the end of the day, the service contract is HTTP and realizing that with C# code is already a mismatch. I know of no wire protocol service descriptions that formally support the concept of inheritance.

Apologies if that seemed like a rant. Stepping of the soapbox...

So far, the most practical away to support this scenario is to use inheritance with re-decoration. This gives you the ability to preserve implementation, but you **must** re-declare what you intend to keep in the current service. It may not be _convenient_, but it's a minimal amount of work with considerable flexibility.

Here's what it would look like:

```c#
[ApiVersion("1.0")]
[RoutePrefix("api/v{version:apiVersion}/orders")]
public class OrdersController : ApiController
{
    [Route("test", Name = "get-v1", Order = 0)]
    public virtual IHttpActionResult Get() => Ok("Get OrdersController v1");

    [Route("test", Name = "delete-v1", Order = 1)]
    public virtual IHttpActionResult Delete(int id) => StatusCode(HttpStatusCode.NoContent);
}

[ApiVersion("2.0")]
[RoutePrefix("api/v{version:apiVersion}/orders")]
public class Orders2Controller : OrdersController
{
    // this method overrides the method defined in v1
    [Route("test", Name = "get-v2", Order = 0)]
    public override IHttpActionResult Get() => Ok("Get Orders2Controller");

    // preserves and exposes the method defined in v1
    [Route("test", Name = "delete-v2", Order = 1)]
    public override IHttpActionResult Delete(int id) => base.Delete(id);
}
```

There are other possibilities as well.  That probably not the answer you were looking or hoping for, but hopefully that provides a few ideas on how to achieve your objectives.

**Comment by doubret** [comment]
Thanks for the alternative solution provided. 

I've always found the fact that actions have to be bound to controllers against nature. Glad to hear that it's gone in .net core. 

I completely understand the way webapi determines a controller/action from a route and the role apiversion plays to desambiguate the targets based on additional version information/constraint.

Looking at the implementation of the DirectRouteControllerSelector it was not clear why the MapToApiVersion shouldn't be taken into consideration when trying to resolve the controller (as the requested version is known and a set of candidate actions have been identified).
I don't have the feeling that it would go against the webapi way of life and in the end it would allow version interleaving across controllers (controllers can declare the same route prefix so they can already be "interleaved" by nature anyway).

When you say that interleaving across types would be quite complex to implement, do you think about something in particular ? Is there more to do than simply adapting the controller selector ?

Maybe I'm missing something here but I fail to see the danger/complexity of adding support for interleaving across types (at first sight it doesn't look a lot more dangerous or complex than what we already have). 

Thanks for your time and feedback.

**Comment by commonsensesoftware** [comment]
Before it seems like I'm resisting, here are a few of the concerns:

* Attributes (e.g. `[MapsToApiVersion]`) isn't the only method to describe versioning. Any solution would have to work with [conventions](../wiki/API-Version-Conventions) as well.
* A service can [_advertise_](../wiki/Version-Advertisement) API versions and discovered API versions are aggregated. _Declared_ refers to the API versions declared on a specific type by decoration or convention.
* Any solution must work in ASP.NET Web API and Core. This might seem unnecessary, but if there isn't strong parity between the two, then the upgrade/migration path is a nightmare. While there are a few subtle differences, the two have strong feature congruence. I've tried to make the APIs as similar and consistent as possible as well. This helps maintain familiarity and source code compatibility when switching platforms. 

It is probably worth describing a few other design details. This is specific to Web API, but the Core version is very similar.

To retrieve the API version model for a service (e.g. controller):
* Query any attributes that implement **IApiVersionProvider**. This is typically ApiVersionAttribute, but can be any attribute. The framework doesn't care about specific attribute implementations.
* Look up the API version model defined by conventions, if any
* Merge API version models as necessary

There is no such thing as an _unversioned_ service/controller. If no attribution or conventions have been applied, then the API version will be based on the configured `DefaultApiVersion` in the options which as the initial value of `1.0`. The only exception is **IApiVersionNeutral** (realized by ApiVersionNeutralAttribute), which indicates that **any** API version is acceptable, including none at all.

Once a controller is matched, the API version models from the other candidates are collated and aggregated into a new API version model that replaces whatever the current controller may have. This is done only once. This process never replaces the _declared_ API versions. This is how the infrastructure disambiguates between all supported and deprecated versions of the service and which versions are directly declared on a type for the resolution process. This is how the values for the `ReportApiVersions` option are collected.

The resolution process for action is much simpler:
* Get _declared_ API versions (a la `[MapToApiVersion]` or conventions)
* If there are no _declared_ API versions, then all _declared_ API versions by the parent controller are implicitly implied

Hopefully that provides some context and you can see that there is some complexity to how the API version information is collected. Based on the Web API design, there would be some additional performance hits to traverse all the actions of all candidates while matching only controllers. Granted, there is likely a design and implementation that can mitigate most of the concerns.

Where things really start to go awry is the mixed routing model in Web API. _Direct Routing_ (aka attribute routing) was bolted onto the existing routing infrastructure. This is where things get really complex. API versioning has to consider both convention-based routes **and** attribute-routes for a given request. It is possible and a supported scenario to have `v1.0` defined with conventions and `v2.0` defined with attributes. This might seem crazy (to us sane devs), but I've seen this in real service codebases. Beyond the esoteric routing mess of a particular service codebase, there are legitimate scenarios for this behavior. A service that transitions to or from OData between API versions would have this problem. Despite OData having it's own attributes for routing, everything is still convention-based under the hood.  This means that using action-level API versions cannot be used purely in the **DirectRouteControllerSelector** or **ConventionRouteControllerSelector**. In fact, the original code in Web API was one source file, but I found it so complex that I separated the two out so I didn't go cross-eyed.

This behavior further continues into the action selection process. If you look at the [ActionSelectorCacheItem.cs](../blob/95756169d6c06660fe8a631c8f9f48b9b8fc3a86/src/Microsoft.AspNet.WebApi.Versioning/Controllers/ActionSelectorCacheItem.cs), you'll see that I changed it (from the original) in several ways. One of the key things is having to support a potential second resolution path to deal with convention vs attribute-based routing.

Again, I'm not entirely opposed to the idea of _implementation inheritance_ to improve developer productivity, but there are a lot subtle considerations. It's starting to sound like I this is a feature request and I should create a new issue to track it specifically. It would likely take a while to implement, but at least there would be some formal tracking and visibility to it.  Feel free to continue sharing your thoughts.

**Comment by doubret** [comment]
I agree that mixing convention based and attribute based routing can be a tricky scenario.
It should be a nice feature though.

I just tested changing the DirectRouteControllerSelector this afternoon and it worked as expected.
Maybe i'll try to make something more advanced based on your advices.

**Comment by commonsensesoftware** [comment]
Making this scenario work is definitely possible - from a technical perspective. I worry about having successful, false-positive tests; it's tricky business. There are a number of tests for mixed routing scenarios. A good example to look at is the [advanced OData acceptance tests](../tree/master/test/Microsoft.AspNet.WebApi.Acceptance.Tests/OData/Advanced).

You might also find it useful to look at the original [DefaultHttpControllerSelector](https://aspnetwebstack.codeplex.com/SourceControl/latest#src/System.Web.Http/Dispatcher/DefaultHttpControllerSelector.cs). It will help you understand why the API versioning works the way it does. I tried not to deviate far from the original to maintain parity.

In retrospect, it seems that API versioning isn't doing a great job at handling scenarios where there is overlap between convention and attribute-based routes for the same API version. I can't remember if I have tests for that. That's a developer mistake, but arguably an edge case. As it stands, it would seem there is not a great experience (at least in Web API) if you have mixed routing and make a mistake.

I need to sit down and really think about how those behaviors would affect evaluation and the developer experience.  Whipping up a simple spec might be useful here. That would also help make sure the behaviors are implemented consistently in ASP.NET Core too, which I'm sure that you'll migrate to eventually. ;)

I'm interested in any details, ideas, etc that you come up with. Hopefully, I've provided enough context to demonstrate the _devil is in the details_. :)

**Comment by dawaji** [comment]
We found a solution and we want your advice about it. We might be missing some points.

By using inheritance and redefining the **DefaultDirectRouteProvider**, we are able to preserve the implementation of the parent without overidding all the older actions. In the case of changing some behavior we need overriding the culprit action.

Here's what it would look like:


```
public class InheritanceDirectRouteProvider : DefaultDirectRouteProvider
    {
        protected override IReadOnlyList<IDirectRouteFactory> 
                         GetActionRouteFactories(HttpActionDescriptor actionDescriptor)
        {
            return actionDescriptor.GetCustomAttributes<IDirectRouteFactory>(inherit: true);
        }
    }
```
 ```
   [RoutePrefix("api/{version:apiversion}/contacts")]
    public class ContactsController : AbstractApiController
    {
        [HttpGet]
        [Route("{contactId:int}")]
        public virtual IHttpActionResult GetContact(int contactId)
     }
```

```
     [ApiVersion("2.0")]
    [RoutePrefix("api/{version:apiversion}/contacts")]
    public class ContactsController : V10.ContactsController
    {
        [HttpGet]
        [Route("{contactId:int}")]
        public override IHttpActionResult GetContact(int contactId)
        {
            }
```

This solution works only with the attribute routing, but in our scenario is enough.

Do you see some flaws in it?

Thanks.

**Comment by commonsensesoftware** [comment]
This seems reasonable. This is kind of the direction I was trying to lead you toward early on. I haven't really experimented with this approach much myself, but a few others have. For your scenario, I think this should be fine.

**Comment by commonsensesoftware** [comment]
It seems as though you are on your way now. I've added issue #230 to formally track your feature request since it has come up several times now. While I might not be in agreement in all scenarios, there are definitely some scenarios where interleaving across types would be beneficial. A scenario that has bothered me for a long time is a DELETE action. No matter many versions of an API you have, this unlikely to ever change in implementation behavior (regardless of whether it's an actual delete, disable, inactivate, or whatever).

If you still have questions or issues, feel free to come back or reopen the issue. Thanks for the discussion. They are always healthy. ;)

**Comment by zdarovka** [comment]
@dawaji Could you please post somewhere working demo? I can not make it work on my project. Also will it work on actions that have are not the overrides but hides the previous implementation ("new" keyword). In my project I need to also change the params that are passed to that action
